homeassistant/components/anomaly/binary_sensor.py

```diff
@@ -0,0 +1,326 @@
+        vol.Optional(CONF_MIN_CHANGE_PERCENT, default=0.0): vol.Coerce(float),
+    }
+)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {vol.Required(CONF_SENSORS): cv.schema_with_slug_keys(SENSOR_SCHEMA)}
+)
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the anomaly sensors."""
+    setup_reload_service(hass, DOMAIN, PLATFORMS)
+
+    sensors = []
+
+    for device_id, device_config in config[CONF_SENSORS].items():
+        entity_id = device_config[ATTR_ENTITY_ID]
+        attribute = device_config.get(CONF_ATTRIBUTE)
+        device_class = device_config.get(CONF_DEVICE_CLASS)
+        friendly_name = device_config.get(ATTR_FRIENDLY_NAME, device_id)
+        invert = device_config[CONF_INVERT]
+        require_both = device_config[CONF_REQUIRE_BOTH]
+        positive_only = device_config[CONF_POSITIVE_ONLY]
+        negative_only = device_config[CONF_NEGATIVE_ONLY]
+        max_samples = device_config[CONF_MAX_SAMPLES]
+        sample_duration = device_config[CONF_SAMPLE_DURATION]
+        max_trailing_samples = device_config[CONF_MAX_TRAILING_SAMPLES]
+        trailing_sample_duration = device_config[CONF_TRAILING_SAMPLE_DURATION]
+        min_change_amount = device_config[CONF_MIN_CHANGE_AMOUNT]
+        min_change_percent = device_config[CONF_MIN_CHANGE_PERCENT]
+
+        sensors.append(
+            SensorAnomaly(
+                hass=hass,
+                device_id=device_id,
+                friendly_name=friendly_name,
+                entity_id=entity_id,
+                attribute=attribute,
+                device_class=device_class,
+                invert=invert,
+                require_both=require_both,
+                positive_only=positive_only,
+                negative_only=negative_only,
+                max_samples=max_samples,
+                sample_duration=sample_duration,
+                max_trailing_samples=max_trailing_samples,
+                trailing_sample_duration=trailing_sample_duration,
+                min_change_amount=min_change_amount,
+                min_change_percent=min_change_percent,
+            )
+        )
+    if sensors:
+        add_entities(sensors)
+
+
+class SensorAnomaly(BinarySensorEntity):
+    """Representation of an anomaly Sensor."""
+
+    def __init__(
+        self,
+        hass,
+        device_id: str,
+        friendly_name: str,
+        entity_id: str,
+        attribute: str,
+        device_class: str,
+        invert: bool,
+        require_both: bool,
+        positive_only: bool,
+        negative_only: bool,
+        max_samples: int,
+        sample_duration: int,
+        max_trailing_samples: int,
+        trailing_sample_duration: int,
+        min_change_amount: float,
+        min_change_percent: float,
+    ):
+        """Initialize the sensor."""
+        self._hass = hass
+        self.entity_id: str = generate_entity_id(ENTITY_ID_FORMAT, device_id, hass=hass)
```

Comment:
Generate the entity id in `async_setup_platform` and pass in the entity_id to the entity instead of `hass`.

Reply:
Is there a sensor you would suggest as  that does all of these things the right way that I can use that as a reference?

Reply:
We normally don't generate the entity_id but just set the name property and let that be a template for the entity_id.

Reply:
I think I got everything right, please let me know if I am doing something wrong on this.